Python: Dataflow, unpacking assignment

yoff opened the pull request:
It looked suspicious that we should need a taint tracking step for this, but it turns out that we had no handling in the data flow layer. I added a test and a data flow step, but we cannot yet remove the taint tracking step without getting test failures.

I think there are at least two reasons for this which should be solved (in future PRs, probably):
1. We have no data flow for slicing
2. We are modelling `TAINTED_LIST` as a variable and should probably use an actually tainted list.


## Update:
- Differences job for latest version: [Job](https://jenkins.internal.semmle.com/job/Changes/job/Python-Differences/366/), no changes, [no change in analysis time](https://jenkins-artifacts.internal.semmle.com/static-files/_2a5ZulasxO8Ui92OWwOTgGvqdcnYqsk-y1qX4foCYUxNjExNDcxMzkwNDcxOjY6cmFzbXVzOmpvYi9DaGFuZ2VzL2pvYi9QeXRob24tRGlmZmVyZW5jZXMvMzY2L2FydGlmYWN0/analysis_time.html)
- Semantics has been simplified by modelling all target lists as tuples.
- Changes from https://github.com/github/codeql/pull/4665 has been merged.

RasmusWL commented:
> We are modelling `TAINTED_LIST` as a variable and should probably use an actually tainted list.

Well. We should probably do both. At least for modeling parts of HTTP requests, the second one is required

```py
my_tainted_list = [TAINTED_STRING, ...]
external_tainted_list = external_lib.returns_tainted_list()
```

RasmusWL reviewed (changes requested):
If we're adding support for iterable unpacking, we need a few more test-cases.

In Python 3, they extended interable unpacking: https://www.python.org/dev/peps/pep-3132/

I created some tests for the python 3 specific stuff in https://github.com/github/codeql/blob/081d66eaa38a911e06c4606ae6c18eb0a55c2c82/python/ql/test/3/library-tests/taint/unpacking/test.py

BUT, notice that this is _iterable_ unpacking. Maybe we should have a `IterableElementContent`?

Some examples we currently wouldn't cover:

```py
a, *b, c = range(10)
print(a, b, c) # 0 [1, 2, 3, 4, 5, 6, 7, 8] 9

a, *b, c = (-i for i in range(10))
print(a, b, c) # 0 [-1, -2, -3, -4, -5, -6, -7, -8] -9

def foo():
    for i in range(10):
        yield i

a, *b, c = foo()
print(a, b, c) # 0 [-1, -2, -3, -4, -5, -6, -7, -8] -9
```

yoff commented:
A warning about reviewing this commit-by-commit. It does add most tests up front and then introduce progressively more complicate predicates to make the tests pass. But then at the end everything is simplified significantly with a big explanatory comment, so it may not be worth it to try to understand the interim state (although it still captures my thinking somewhat). I also got the test annotations wrong initially (and a few times later), so that might be confusing.

Let me know if you want me to produce a polished history..

tausbn commented on python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:
```diff
@@ -1008,6 +1028,207 @@ predicate subscriptReadStep(CfgNode nodeFrom, Content c, CfgNode nodeTo) {
+    UnpackingAssignmentTarget() {
+      this instanceof UnpackingAssignmentDirectTarget
+      or
+      exists(UnpackingAssignmentTarget parent | this = parent.getAnElement())
```
```suggestion
      this = any(UnpackingAssignmentTarget parent).getAnElement()
```

tausbn commented on python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:
```diff
@@ -1008,6 +1028,207 @@ predicate subscriptReadStep(CfgNode nodeFrom, Content c, CfgNode nodeTo) {
+  }
+
+  /** A (possibly recursive) target of an unpacking assignment */
+  class UnpackingAssignmentTarget extends ControlFlowNode {
```
Could this extend `SequenceNode` instead (in which case we can drop the cast in `getElement`)?

yoff replied:
No, because not all targets are sequences. For instance in `a, (b, c) = iterable`, only `a, (b, c)` and (a, b) are sequence, while `a`, `b`, and `c` ar not.

tausbn replied:
Ah, good point. I was misled by the fact that the only place where it _isn't_ a `SequenceNode` is in the definition of `TIterableElement`.

Perhaps it would be better to split this out into two subclasses, one for the "leaves" and one for the internal sequence nodes? That way, we would also avoid the slight code smell of `getElement/1` silently failing at the leaves. And we could get rid of all of the `instanceof SequenceNode` guards.

(Or perhaps we don't even need the "leaves only" class, since we never actually do things with _just_ those nodes. A simple subclass that represents `UnpackingAssignmentSequenceTarget`s should suffice.)

yoff replied:
`UnpackingAssignmentSequenceTarget` works nicely :-)

tausbn commented on python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:
```diff
@@ -1008,6 +1028,207 @@ predicate subscriptReadStep(CfgNode nodeFrom, Content c, CfgNode nodeTo) {
+  }
+}
+
+import unpackinAssignment
```
```suggestion
import UnpackingAssignment
```

RasmusWL reviewed (changes requested):
I've not reviewed all code in full yet, actually just the tests, but I'm going to request changes now, and go through the rest of the PR.

### Iterable unpacking in `for`

We don't have any tests for using iterable unpacking in `for`, for example

```
@expects(2)
def test_iterable_unpacking_in_for():
    tl = [(SOURCE, NONSOURCE), (SOURCE, NONSOURCE)]
    for x,y in tl:
        SINK(x)
        SINK_F(y)
```

but it would also be nice to have one that uses new extended `*` feature from Python 3.

tausbn commented on python/ql/test/experimental/dataflow/coverage/test.py:
```diff
@@ -516,6 +516,110 @@ def test_swap():
+    # tuple
+    ((a1, a2, a3), b, c) = ll
+    SINK(a1)
+    SINK_F(a2)  # We expect an FP as all elements are tainted
```
This bothers me a bit. It feels slightly wrong to use "any element can be tainted" when we actually have enough information to reconstruct which elements are in fact tainted.

Perhaps we could do something along these lines:

- Track list indices precisely for the cases where are able, _but_
- have this gracefully devolve into "some element is tainted" when we perform operations that may move things around (so something like `list.append` is probably fine, as it only adds to the end of the list. `list.sort` would go from specific to generic taint, etc.)

This should probably not happen in this PR, though.

yoff replied:
I would be happy to try this out at some point. Note, though, that we are already more precise than C# where tuple elements are modelled as `ElementContent`, so we may or may not need al this precision. I agree that introducing precise element information and having a graceful mutation to imprecise information is interesting, not only for lists. But I do think it is out of scope for this PR.

tausbn commented on python/ql/src/semmle/python/dataflow/new/internal/DataFlowPublic.qll:
```diff
@@ -319,6 +328,42 @@ class KwUnpacked extends Node, TKwUnpacked {
+
+/**
+ * A synthetic node representing an iterable element. Used for changing content type
+ * for instance from a `ListElement` to a `TupleElement`.
```
This example is exactly the same as for `IterableSequence`, which may make it unclear why both classes are needed.

yoff replied:
Right, I could add something like "via a read step to an IterableElement followed by a store step" to the above and "usually via a read step from an IterableSequence followed by a store step" here.

tausbn reviewed (changes requested):
Still chewing over some of the finer details of this, but I figured this was a good point to submit my review.

yoff commented on python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:
```diff
@@ -126,6 +126,20 @@ module EssaFlow {
+    //   `[a, b] = iterable`
+    //   nodeFrom = `iterable`, cfg node
+    //   nodeTo = `TIterableSequence([a, b])`
+    exists(UnpackingAssignmentDirectTarget target |
+      nodeFrom.asExpr() = target.getValue() and
+      nodeTo = TIterableSequence(target)
+    )
+    or
+    exists(Assign assign, SequenceNode target | target.getNode() = assign.getATarget() |
+      nodeFrom.asExpr() = assign.getValue() and
+      nodeTo.asCfgNode() = target
+    )
```
I would think that `nodeTo` being of branch `TCfgNode` is incompatible with it being of branch `TIterableSequence`. Or do you mean more disjoint than that?

yoff replied:
Actually, the second case should just be removed (I was prompted by `@RasmusWL`'s comment to look more closely).

yoff commented on python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:
```diff
@@ -1008,6 +1028,207 @@ predicate subscriptReadStep(CfgNode nodeFrom, Content c, CfgNode nodeTo) {
+ * In order to preserve precise content, we also take a flow step from `TIterableSequence(receiver)`
+ * directly to `receiver`.
+ *
+ * The strategy is then via several read-, store-, and flow steps:
```
Yes, that is a good idea and should not be difficult to add.

yoff replied:
I opted for an example after the step; inlining did not work quite as well.

tausbn commented on python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:
```diff
@@ -1008,6 +1028,207 @@ predicate subscriptReadStep(CfgNode nodeFrom, Content c, CfgNode nodeTo) {
+          nodeTo = TIterableSequence(element) and
+          precise = true
+        else
+          if element.getNode() instanceof Starred
```
Looking at this and comparing it with line 1191, I was wondering if these could actually overlap (and so whether a simple  disjunction would suffice). Then I realised that the following abomination is valid:
```py
a, *(b, c) = (1,2,3)
```
Do you know if we handle the above correctly?

RasmusWL replied:
yikes :smile:

yoff replied:
I would expect that we do not...

yoff replied:
Test added, we do indeed not.

RasmusWL reviewed (commented):
Overall I think these new complex tests show that it would be _really_ nice to have inline-test-expectations for these tests. I would almost say it's essential to get done soon, since I did not look at the change to the .expected file, and therefore don't actually know what things this PR handles and not handles :(

I've added some comments here requesting to get things cleared up, but would like to chat with you (live) when you've had a chance to look them over, so I can better understand the whole of this PR :wink: just ping me when you're ready to do so

yoff commented on python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:
```diff
@@ -126,6 +126,20 @@ module EssaFlow {
+      nodeFrom.asExpr() = target.getValue() and
+      nodeTo = TIterableSequence(target)
+    )
+    or
+    exists(Assign assign, SequenceNode target | target.getNode() = assign.getATarget() |
+      nodeFrom.asExpr() = assign.getValue() and
+      nodeTo.asCfgNode() = target
+    )
```
Good catch! I think this case is a left-over that should be deleted.

yoff commented on python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:
```diff
@@ -161,6 +175,8 @@ module EssaFlow {
     // If expressions
     nodeFrom.asCfgNode() = nodeTo.asCfgNode().(IfExprNode).getAnOperand()
     or
+    unpackingAssignmentFlowStep(nodeFrom, nodeTo)
```
A short one, perhaps? I think the documentation is on the module, there should also be a comment on the predicate itself

yoff replied:
I managed to make a short one that actually has some information :-)

yoff commented on python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:
````diff
@@ -1008,6 +1028,207 @@ predicate subscriptReadStep(CfgNode nodeFrom, Content c, CfgNode nodeTo) {
+ *
+ * Using wildcards we may have
+ * ```python
+ *   (a, *b) = ("a", "b", "tainted string")  # RHS has content `TupleElement(2)`
````
Thanks, it should be written out in full everywhere.
`"tainted string"` should perhaps be replaced with `SOURCE`.

yoff commented on python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:
````diff
@@ -1008,6 +1028,207 @@ predicate subscriptReadStep(CfgNode nodeFrom, Content c, CfgNode nodeTo) {
+ * ```python
+ *   (a, [b, *c]) = ("a", ("tainted string", "c"))  # RHS has content `TupleElement(1); TupleElement(0)`
+ * ```
+ * where `a` should not receive content, but `b` and `c` should. `c` will be `["c"]` so
+ * should have the content converted and transferred, while `b` should read it.
````
In this case `b` and `c` are both inside a list so if there is any content in the tuple matched by `[b, *c]` then `[b, *c]` should have content `ListElement [rest of acces path]` and `b` should have `[rest of acces path]` and `c` should have `ListElement [rest of acces path]`.

RasmusWL replied:
Hmm, my understanding of iterable unpacking is that it doesn't matter if you use normal parentheses or square brackets to match the pattern (but that having both can enable writing clean code). So

```py
a, (b, c) = iterable
(a, (b, c)) = iterable
[a, [b, c]] = iterable
a, [b, c] = iterable
```

are all semantically equivalent -- or have I misunderstood something?

yoff replied:
I would say that, apart from the first two, these are all different. But the difference is only in the types of intermediate results, so you only see it if that difference matters. It might for instance matter for runtime. It matters for us, because we distinguish list-content and tuple-content. In terms of the values of `a`, `b`, and `c`, the interpreter does not see a difference because the third element of a list is just as precisely known as the third element of a tuple and so it can move back and forth between the two without noticing. Our analysis does see a difference, which is why it does not agree with the interpreter on the following program:
```python
  l = [SOURCE, NONSOURCE]
  t = (l[0], l[1])
```
The interpreter sees `t` having `SOURCE` only in the first component, our analysis says it can be found in both.

RasmusWL replied:
So, just to be perfectly clear, do you agree that these are all semantically equivalent? (in terms of Python semantics)

yoff replied:
What is Python semantics? They may incur different runtimes or memory footprints, and if you manage to change the subscripting of either list or tuple, they could also give different values.

RasmusWL replied:
What is anything, really? :)

My point here is that we could make our analysis a bit simpler, by ignoring whether square brackets or normal parenthesis was used on the LHS of the assignment. I'm guessing that with our current Python libraries, we represent these as `TupleNode` or `ListNode`, but I don't think we need to make the distinction here.

After spending some time looking for evidence on this, I finally found https://docs.python.org/3/reference/simple_stmts.html#assignment-statements:

> Assignment of an object to a target list, optionally enclosed in parentheses or square brackets, is recursively defined as follows.

So if a target is actually a target list (to use the wording from that Python reference), I think we can just consider that target list to be a tuple. This will allow us to simplify the data-flow modeling a bit, and might lead to a bit more precision in some cases (for example `(a, [b, *c]) = ("a", ("tainted string", "c"))`).

### non-goals

I'm not suggesting we should change the behavior of our data-flow modeling in general. So in the example below, when tracking data-flow from `SOURCE`, I'm perfectly fine with modeling content of `l` as _some element_, meaning that we think the content might flow to `b`. If we ever want to change that, this is not the PR to do it in at least :smile: 

```py
l = [SOURCE, NON_SOURCE]
t = (SOURCE, NON_SOURCE)

a, b = l
x, y = t
```

yoff replied:
> What is anything, really? :)

I meant that one could define many semantics for Python, depending on ones interest. One such might or might not include expected runtime or memory footprint.

> After spending some time looking for evidence on this, I finally found https://docs.python.org/3/reference/simple_stmts.html#assignment-statements:
> 
> > Assignment of an object to a target list, optionally enclosed in parentheses or square brackets, is recursively defined as follows.
> 
> So if a target is actually a target list (to use the wording from that Python reference), I think we can just consider that target list to be a tuple. This will allow us to simplify the data-flow modeling a bit, and might lead to a bit more precision in some cases (for example `(a, [b, *c]) = ("a", ("tainted string", "c"))`).

Indeed, it looks like you are right! The different notations are just there for convenience and always denote a "target list", which can indeed model as a tuple.

Reading the reference, it also occurred to me that we may not handle the assignment (and the possible call to a special method) hidden in
```python
  a, b[0] = 1, 2
```

RasmusWL replied:
```py
 a, b[0] = 1, 2
```

\*goes to check in ipython\* ... good catch! Would be great with a test-case at least, but not sure if you want to handle that (more general) case in this PR?

tausbn replied:
I think adding a test case is probably a good idea, but I don't think fixing this needs to be done in this PR.
Also, it seems like this is two different issues:

1. Correct handling of content when assigning to a subscripted expression, and
2. correctly handling the case where `__setitem__` exists on `b`.

Of these, I would say 1. is probably the higher priority (though there may exist fancy libraries that use `__setitem__` to do clever stuff, so we may end up needing both).

yoff commented on python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:
```diff
@@ -1008,6 +1028,207 @@ predicate subscriptReadStep(CfgNode nodeFrom, Content c, CfgNode nodeTo) {
+ * directly to `receiver`.
+ *
+ * The strategy is then via several read-, store-, and flow steps:
+ * 1. [Flow] Content is transferred from `iterable` to `TIterableSequence(sequence)` via a
+ *    flow step. From here, everything happens on the LHS.
+ *
+ * 1. [Flow] Content is transferred from `TIterableSequence(sequence)` to `sequence` via a
+ *    flow step.
+ *
+ * 1. [Read] Content is read from `TIterableSequence(sequence)` into  `TIterableElement(sequence)`.
+ *    If `sequence` is of type tuple, we will not read tuple content as that would allow
+ *    cross talk.
+ *
+ * 1. [Store] Content is stored from `TIterableElement(sequence)` to `sequence`.
+ *    Here the content type is chosen according to the type of sequence.
+ *
+ * 1. [Read] Content is read from `sequence` to its elements according to the type of `sequence`.
+ *    If the element is a plain variable, the target is the corresponding essa node.
+ *    If the element is itelf a sequence, with control-flow node `seq`, the target is `TIterableSequence(seq)`.
+ *    If the element is a starred variable, with control-flow node `v`, the target is `TIterableElement(v)`.
+ *
+ * 1. [Store] Content is stored from `TIterableElement(v)` to the essa variable for `v`, with
+ *    content type `ListElement`.
+ *
+ * 1. [Flow, Read, Store] The last 5 steps are repeated for all recursive elements which are sequences.
```
They get real numbers in the pop-up, and we are not referring to the numbers anyway. I guess I read the markdown more than I read the pop-up, though, so perhaps it would be helpful..

yoff replied:
Ok, we might refer to them from comments on the predicates.

yoff commented on python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:
```diff
@@ -1008,6 +1028,207 @@ predicate subscriptReadStep(CfgNode nodeFrom, Content c, CfgNode nodeTo) {
+    ControlFlowNode getAnElement() { result = this.getElement(_) }
+  }
+
+  predicate unpackingAssignmentFlowStep(Node nodeFrom, Node nodeTo) {
```
That is a good idea.

tausbn replied:
While I like the link back to the steps as outlined above (and the description one can find there), I think I would prefer it if it was possible to understand this predicate _in situ_. That is, some documentation along the lines of "holds if ... flow from `nodeFrom` to `nodeTo`". When seeing a _use_ of the predicate, "`Step 2`" is unlikely to be very illuminating in the resulting intelliSense.

yoff replied:
Fair point.

yoff commented on python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:
```diff
@@ -1008,6 +1028,207 @@ predicate subscriptReadStep(CfgNode nodeFrom, Content c, CfgNode nodeTo) {
+        exists(int index | exists(target.getElement(index)) |
+          c.(TupleElementContent).getIndex() = index
+        )
+        // leaving out dict content for now
```
The thing is, it would unpack the keys, so I am not sure how to handle that. But I agree that we should either make it a TODO or write in a more prominent place that we are not handling it.

yoff commented:
I have addressed the issues raised so far except `for`. Mentally I have been solving unpacking assignment rather than iterated unpacking. I realise, that we should probably rename the module and have it be used in all the contexts where iterated unpacking happen (`for`-iteration and comprehensions). And it probably will not be very much code to add `for` and I need to add a `for` step anyway, so I will try that shortly. But those extensions could also be postponed to a separate PR...

tausbn commented on python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:
```diff
@@ -1008,6 +1028,270 @@ predicate subscriptReadStep(CfgNode nodeFrom, Content c, CfgNode nodeTo) {
+  }
+
+  /** A (possibly recursive) target of an unpacking assignment which is also a sequence. */
+  class UnpackingAssignmentSequenceTarget extends UnpackingAssignmentTarget {
+    UnpackingAssignmentSequenceTarget() { this instanceof SequenceNode }
+  }
```
Could just also extend `SequenceNode` and get rid of the charpred entirely.

yoff replied:
No, `SequenceNode` is abstract.

tausbn replied:
Ah, of course it is. 👍

tausbn commented on python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:
```diff
@@ -1008,6 +1028,270 @@ predicate subscriptReadStep(CfgNode nodeFrom, Content c, CfgNode nodeTo) {
+    UnpackingAssignmentTarget() {
+      this instanceof UnpackingAssignmentDirectTarget
+      or
+      this = any(UnpackingAssignmentTarget parent).getAnElement()
+    }
+
+    ControlFlowNode getElement(int i) { result = this.(SequenceNode).getElement(i) }
+
+    ControlFlowNode getAnElement() { result = this.getElement(_) }
```
If you change the first line to `this = any(UnpackingAssignmentSequenceTarget parent).getAnElement()`, then you can maybe move the two methods into `UnpackingAssignmentSequenceTarget` where they make more sense? (And then the cast is unnecessary.)

yoff replied:
Yes, that should be nicer 👍

tausbn commented on python/ql/src/semmle/python/dataflow/new/internal/DataFlowPublic.qll:
```diff
@@ -60,7 +60,16 @@ newtype TNode =
+  /**
+   * A synthetic node representing that an iterable sequence flows to consumer.
+   */
+  TIterableSequence(UnpackingAssignmentTarget consumer) { consumer instanceof SequenceNode } or
```
```suggestion
  TIterableSequence(UnpackingAssignmentSequenceTarget consumer) or
```
(May need additional reformatting.)

tausbn commented on python/ql/src/semmle/python/dataflow/new/internal/DataFlowPublic.qll:
```diff
@@ -60,7 +60,16 @@ newtype TNode =
+   * A synthetic node representing that there may be an iterable element
+   * for `consumer` to consume.
+   */
+  TIterableElement(UnpackingAssignmentTarget consumer)
```
For most of these disjuncts, we have names of the form `...Node` (though notably `TKwUnpacked` slipped through the cracks somehow). Should we perhaps endeavour to keep this consistent?

yoff replied:
Sure, that does seem to be a convention :-)

tausbn commented on python/ql/src/semmle/python/dataflow/new/internal/DataFlowPublic.qll:
```diff
@@ -319,6 +328,46 @@ class KwUnpacked extends Node, TKwUnpacked {
+  override string toString() { result = "IterableSequence" }
+
+  override DataFlowCallable getEnclosingCallable() {
+    result = any(CfgNode node | node = TCfgNode(consumer)).getEnclosingCallable()
```
There's something about the way this is written that bugs me a bit. I'm guessing it is the case that any `IterableSequence` is in fact also a `CfgNode` (otherwise we would have a consistency problem for this predicate). In that case, it's slightly awkward that we have to jump back through the IPA type constructor to get this link back.

Could we perhaps instead just say that `TIterableSequence` contains not just a `ControlFlowNode`, but in fact a `CfgNode` that itself contains said `ControlFlowNode`?

This might make other parts of the code a bit more awkward, though...

yoff replied:
Perhaps we could do a more local refactor, making the field, `consumer`, a `CfgNode` and have the charpred link them up?

tausbn reviewed (changes requested):
I think this wins the prize for "biggest QLDoc comment". :smile: 
I have added a few comments and suggestions, but I think this looks like it should be possible to merge really soon. Nicely done! :tada:

yoff commented:
Differences job for version without unhelpful store steps: [Job](https://jenkins.internal.semmle.com/job/Changes/job/Python-Differences/359/) showing no changes and [analysis times](https://jenkins-artifacts.internal.semmle.com/static-files/cZ-xlcl3Kxq36GEwYGfXJjuC982kg7OLFZ89vEGpjsMxNjExMDgxNTEzNTY0OjY6cmFzbXVzOmpvYi9DaGFuZ2VzL2pvYi9QeXRob24tRGlmZmVyZW5jZXMvMzU5L2FydGlmYWN0/analysis_time.html) with salt seeing an increase just shy of 10% (so somewhat concerning).

yoff commented:
New job after merging in main: [Job](https://jenkins.internal.semmle.com/job/Changes/job/Python-Differences/361/)

yoff commented:
Simplified the model, knowing that all LHS sequences are the same (now modelled with `TupleElementContent`). It did not immediately allow me to remove `TIterableSequence` as I had hoped, because we may still read some nested list content that we need to converge.

I also added slightly more precise modelling of indices in the presence of starred variables.

yoff commented:
New [job](https://jenkins.internal.semmle.com/job/Changes/job/Python-Differences/366/).

yoff commented:
Invalid test was rejected (yay).

tausbn commented on python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:
```diff
@@ -995,6 +1017,289 @@ predicate subscriptReadStep(CfgNode nodeFrom, Content c, CfgNode nodeTo) {
+ * will need to change type if it should be transferred from the LHS to the RHS.
+ *
+ * Note that (CodeQL modeling of) content does not have to change type on data-flow
+ * path _inside_ the LHS, as the different allowed syntaxes here are merely a convenience.
```
```suggestion
 * paths _inside_ the LHS, as the different allowed syntaxes here are merely a convenience.
```

tausbn commented on python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:
```diff
@@ -995,6 +1017,289 @@ predicate subscriptReadStep(CfgNode nodeFrom, Content c, CfgNode nodeTo) {
+ *    flow step.
+ *
+ * 3. [Read] Content is read from `TIterableSequence(sequence)` into  `TIterableElement(sequence)`.
+ *    As `sequence` is modelled as a tuple, we will not read tuple content as that would allow
```
```suggestion
 *    As `sequence` is modeled as a tuple, we will not read tuple content as that would allow
```

tausbn commented on python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:
```diff
@@ -995,6 +1017,289 @@ predicate subscriptReadStep(CfgNode nodeFrom, Content c, CfgNode nodeTo) {
+ *
+ * 3. [Read] Content is read from `TIterableSequence(sequence)` into  `TIterableElement(sequence)`.
+ *    As `sequence` is modelled as a tuple, we will not read tuple content as that would allow
+ *    cross talk.
```
```suggestion
 *    crosstalk.
```

tausbn commented on python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:
```diff
@@ -995,6 +1017,289 @@ predicate subscriptReadStep(CfgNode nodeFrom, Content c, CfgNode nodeTo) {
+ * 5. [Read] Content is read from `sequence` to its elements.
+ *    a) If the element is a plain variable, the target is the corresponding essa node.
+ *
+ *    b) If the element is itelf a sequence, with control-flow node `seq`, the target is `TIterableSequence(seq)`.
```
```suggestion
 *    b) If the element is itself a sequence, with control-flow node `seq`, the target is `TIterableSequence(seq)`.
```

tausbn commented on python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:
```diff
@@ -995,6 +1017,289 @@ predicate subscriptReadStep(CfgNode nodeFrom, Content c, CfgNode nodeTo) {
+  /**
+   * Step 3
+   * Data flows from `TIterableSequence(sequence)` into  `TIterableElement(sequence)`.
+   * As `sequence` is modelled as a tuple, we will not read tuple content as that would allow
+   * cross talk.
```
```suggestion
   * As `sequence` is modeled as a tuple, we will not read tuple content as that would allow
   * crosstalk.
```

tausbn commented on python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:
```diff
@@ -995,6 +1017,289 @@ predicate subscriptReadStep(CfgNode nodeFrom, Content c, CfgNode nodeTo) {
+   * three cases for what `toNode` should be:
+   *    a) If the element is a plain variable, `toNode` is the corresponding essa node.
+   *
+   *    b) If the element is itelf a sequence, with control-flow node `seq`, `toNode` is `TIterableSequence(seq)`.
```
```suggestion
   *    b) If the element is itself a sequence, with control-flow node `seq`, `toNode` is `TIterableSequence(seq)`.
```

tausbn reviewed (changes requested):
A few minor bits, but then I'll merge it, I promise! :)

RasmusWL commented on python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:
```diff
@@ -995,6 +1017,289 @@ predicate subscriptReadStep(CfgNode nodeFrom, Content c, CfgNode nodeTo) {
+ * where `a` should not receive content, but `b` and `c` should. `c` will be `[SOURCE]` so
+ * should have the content converted and transferred, while `b` should read it.
+ *
+ * The strategy for converting content type is to break the transfer up into a read step
+ * and a store step, together creating a converting transfer step.
+ * For this we need a synthetic node in the middle, which we call `TIterableElement(receiver)`.
+ * It is associated with the receiver of the transfer, because we know the receiver type (tuple) from the syntax.
+ * Since we sometimes need a converting read step (in the example above, `[b, *c]` reads the content
+ * `ListElementContent` but should have content `TupleElementContent(0)` and `TupleElementContent(0)`),
+ * we actually need a second synthetic node. A converting read step is a read step followed by a
+ * converting transfer.
+ *
+ * We can have a uniform treatment by always having two synthetic nodes and so we can view it as
+ * two stages of the same node. So we read into (or transfer to) `TIterableSequence(receiver)`,
+ * from which we take a read step to `TIterableElement(receiver)` and then a store step to `receiver`.
+ *
+ * In order to preserve precise content, we also take a flow step from `TIterableSequence(receiver)`
+ * directly to `receiver`.
+ *
+ * The strategy is then via several read-, store-, and flow steps:
```
I think this part is unclear, and I would like to rewrite it.

```suggestion
 * To transfer content from RHS to the elements of the LHS in the expression `sequence = iterable`, we use two synthetic nodes:
 * - `TIterableSequence(sequence)` which captures the content-modeling the entire `sequence` will have 
 * (essentially just a copy of the content-modeling the RHS has)
 * - `TIterableElement(sequence)` which captures the content-modeling that will be assigned to an element. Note that
 * an empty access path means that the value we are tracking flows directly to the element.
 * 
 * Since we need to handle recursive structures on the LHS, we can have a uniform treatment by always having these 
 * two synthetic nodes and so we can view it as two stages of the same node. So we read into (or transfer to) 
 * `TIterableSequence(receiver)`, from which we take a read step to `TIterableElement(receiver)` and then a store step 
 * to `receiver`.
 *
 * This is accomplished via several read-, store-, and flow steps:
```

yoff replied:
Did you mean to leave out this bit "In order to preserve precise content, we also take a flow step from `TIterableSequence(receiver)` directly to `receiver`."? (Do you feel that is implied?)
Also, do you feel the two stages help? I originally wanted to make it a single branch `TIterableElement(receiver, direction)` with `direction` being `"in"` or `"out"`, but the two branches seemed much clearer end cleaner. So perhaps the two stages idea is not relevant anymore?

RasmusWL replied:
> * Did you mean to leave out this bit "In order to preserve precise content, we also take a flow step from `TIterableSequence(receiver)` directly to `receiver`."? (Do you feel that is implied?)

I must admit that "In order to preserve precise content, we also take a flow step from `TIterableSequence(receiver)` directly to `receiver`." never really made sense to me.

My understanding is that the only time _Step 2_ (`TIterableSequence(sequence)` -> `sequence`) is useful, is when the content-modeling is already `TupleElementContent`. The `unpackingAssignmentFlowStep` predicate doesn't have any constraint showing this, but the only way to transfer things to the elements of the sequence is to use _Step 5_, which always requires that the content-modeling is `TupleElementContent`.

If I'm misunderstanding somewhere here, we definitely need to highlight this in some way. If my understanding is correct, we might want to add a comment to the explanation of _Step 2_, saying it's only needed to handle things (from RHS) where the content-modeling is already using `TupleElementContent`

> * Since we are here, do you feel the two stages help? I originally wanted to make it a single branch `TIterableElement(receiver, direction)` with `direction` being `"in"` or `"out"`, but the two branches seemed much clearer end cleaner. So perhaps the two stages idea is not relevant anymore?

No I'm happy about the current setup :+1:

yoff replied:
I ended up creating a mashup, turns out I still felt a need for justifying `TIterableSequence` 😅
Your understanding above is correct and hopefully now fully reflected in the comment.

RasmusWL commented on python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:
```diff
@@ -995,6 +1017,289 @@ predicate subscriptReadStep(CfgNode nodeFrom, Content c, CfgNode nodeTo) {
+ * 7. [Flow, Read, Store] Steps 2 through 7 are repeated for all recursive elements which are sequences.
+ *
+ *
+ * We illustrate the above steps on the assignment
```
```suggestion
 * We illustrate the above steps on the assignment
 * 
 * ```python
 * (a, b) = ["a", SOURCE]
 * ```
 *
 * Looking at the content propagation to `a`:
 *   `["a", SOURCE]`: [ListElementContent]
 *
 * --Step 1-->
 * 
 *   `TIterableSequence((a, b))`: [ListElementContent]
 *
 * --Step 3-->
 * 
 *   `TIterableElement((a, b))`: []
 *
 * --Step 4-->
 * 
 *   `(a, b)`: [TupleElementContent(0)]
 *
 * --Step 5c-->
 *
 *   `a`: []
 * 
 * Meaning there is data-flow from the RHS to `a` (an over approximation). The same logic is applied to show there is data-flow to `b`. Note that _Step 3_ and _Step 4_ would not have been needed if the RHS had been a tuple.
 * 
 * Another, more complicated example:
```

RasmusWL replied:
I didn't actually run this example, I only mentally walked through the process.

RasmusWL reviewed (changes requested):
I spend a considerable amount of time looking through the comments, and trying to understand it all :) I think I do now, and the docs were very helpful :muscle:

I've made a couple of suggestions to make them even better (I hope)

yoff commented on python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:
```diff
@@ -995,6 +1017,316 @@ predicate subscriptReadStep(CfgNode nodeFrom, Content c, CfgNode nodeTo) {
+ * 
+ *   `(a, b)`: [TupleElementContent(0)]
+ *
+ * --Step 5c-->
```
```suggestion
 * --Step 5a-->
```

RasmusWL reviewed (approved):
Thanks for the clarifications. I think it will greatly help anyone who dares to look at this in the future :smile: ... probably to the point where it doesn't feel so complicated that it can be described as huge undertaking to look at it (since the docs are more clear now, it's easier to understand) :muscle:

tausbn reviewed (approved):
:rocket: :tada: 

What a ride!
